blogposts/2022/css-modules-migration.md

```diff
@@ -0,0 +1,210 @@
+
+### 2. Convert global styles file into a CSS module.
+
+It’s a multistep operation:
+
+- First, get rid of the BEM notation used for CSS selectors and remove redundant nesting required to ensure selector uniqueness.
+
+```css
+.insights-dashboard {
+  flex: none;
+
+  &__wrapper {
+    display: flex;
+  }
+}
+```
+
+⬇
+
+```css
+.insights-dashboard {
+  flex: none;
+}
+
+.wrapper {
+  display: flex;
+}
+```
+
+- To achieve this, we used [PostCSS](https://github.com/postcss/postcss)—a ​tool for transforming styles with JS plugins. The resulting script is similar to [the postcss-nested plugin](https://github.com/postcss/postcss-nested), with some adjustments to remove the BEM notation.
+- Save a list of CSS Module classes in memory for later reference in the React component. It is done by using utility functions provided by [the css-modules-loader-core package](https://github.com/css-modules/css-modules-loader-core). They give us a list of classes available in a CSS Module that we can map ourselves to global CSS classes that we just transformed. For our toy example, it would look like this:
+
+```css
+const classMapping = {
+  'insights-dashboard':'insights-dashboard','insights-dashboard__wrapper' : 'wrapper';
+}
+```
+
+- As a final touch, add `.module.scss` to the filename required for build-tools to interpret this file as a CSS Module.
+
+### 3. Replace references to global classes in the React component using a list of CSS Modules classes preserved from the previous step.
+
+To manipulate Typescript AST, we used [ts-morph](https://github.com/dsherret/ts-morph) —TypeScript Compiler API wrapper for static analysis and programmatic code changes. Relying on this package API, the codemod script iterates over all string literals in the React component AST and searches for global classes processed in the previous step.
+
+Here’s the AST generated for our small example. Explore it yourself using [astexplorer](https://astexplorer.net/#/gist/eca630c9f5464e6b027ec13aac91711a/4295e325b8cdad474f9d4f559248f534e709356e).
+
+<figure>
+  <img src="https://storage.googleapis.com/sourcegraph-assets/blog/code-insights-ga-blogs/insights-dashboard-ast.png" alt="Insights dashboard AST"
+</figure>
+
+Codemod replaces every string literal match with reference to the corresponding CSS Module class.
+
+```css
+<div className="insights-dashboard__wrapper" />
+```
+
+⬇
+
+```css
+<div className={styles.wrapper} />
+```
+
+At this point, the transformation is complete, and the script outputs information about CSS classes not used in the React component. It allowed us to remove dead code in the migration process.
+
+The codemod script can be used as a CLI tool by supplying a list of files to transform:
+
+`$ yarn transform --write --transform ./globalCssModule.ts ./sourcegraph/**/*.tsx`
```

Comment:
Was trying to capture this, need to check if correct
![image1](https://user-images.githubusercontent.com/82806852/155334361-dfdef497-e99a-44fb-8f90-ae6353b179e7.png)

Reply:
```suggestion
`$ yarn transform --write --transform ./globalCssToCssModule.ts ./sourcegraph/**/*.tsx`
```

Reply:
Applied in batched suggestions